lib/portage_utils.py

```diff
@@ -152,38 +160,42 @@ def open(filename, mode='r', quiet=True):
    filename: name of the file to open
    mode: open mode
    quiet:  suppress "zcat: stdout: Broken pipe" messages.
+   encoding: defaults to "utf8" for text modes
    return: file handle to the open file.
    """
+   if encoding is None and mode not in ("rb", "wb"):
+      encoding="utf8"
+
    filename.strip()
    #debug("open: ", filename, " in ", mode, " mode")
-   if len(filename) is 0:
+   if len(filename) == 0:
       fatal_error("You must provide a filename")
 
    if filename == "-":
-      if mode == 'r':
+      if mode in ('r', 'rt'):
          theFile = sys.stdin
-      elif mode == 'w':
+      elif mode in ('w', 'wt'):
          theFile = sys.stdout
       else:
          fatal_error("Unsupported mode.")
    elif filename.endswith('|'):
-      theFile = Popen(filename[:-1], shell=True, executable="/bin/bash", stdout=PIPE).stdout
+      theFile = Popen(filename[:-1], shell=True, executable="/bin/bash", encoding=encoding, stdout=PIPE).stdout
    elif filename.startswith('|'):
-      theFile = Popen(filename[1:], shell=True, executable="/bin/bash", stdin=PIPE).stdin
+      theFile = Popen(filename[1:], shell=True, executable="/bin/bash", encoding=encoding, stdin=PIPE).stdin
    elif filename.endswith(".gz"):
       #theFile = gzip.open(filename, mode+'b')
-      if mode == 'r':
+      if mode in ('r', 'rt', 'rb'):
          if quiet:
-            theFile = Popen(["zcat", "-f", filename], stdout=PIPE, stderr=open('/dev/null', 'w')).stdout
+            theFile = Popen(["zcat", "-f", filename], stdout=PIPE, encoding=encoding, stderr=open('/dev/null', 'w')).stdout
          else:
-            theFile = Popen(["zcat", "-f", filename], stdout=PIPE).stdout
-      elif mode == 'w':
-         internal_file = builtins.open(filename, mode)
-         theFile = Popen(["gzip"], close_fds=True, stdin=PIPE, stdout=internal_file).stdin
+            theFile = Popen(["zcat", "-f", filename], stdout=PIPE, encoding=encoding).stdout
+      elif mode in ('w', 'wt', 'rb'):
```

Comment:
Oh yeah, that's definitely a typo!

Reply:
I'm not actually sure the binary mode works properly when we're opening pipes or gzipped files.

Reply:
And I think the append mode would be hard to support correctly with those, as well. In the original code, Darlene had a fatal_error for anything but `r` and `w`. Maybe I should go back to that, because I'm not sure what I'm doing is correct in all cases. And our unit tests don't cover any cases other that `r` and `w`.

Reply:
Let's not do the append mode because I'm not sure it is a use case that we have.

Reply:
Looking at the code, `internal_file` is the output of the `gzip` pipe, so it's actually binary. My code works anyway, probably because `Popen(stdout=internal_file)` effectives does `write()` calls to `internal_file`, and never `print()` ones, so the `t` vs `b` part of its mode is just ignored.

Reply:
I take it back. I just tested, and for reasons I don't understand, `open("delme.gz", mode="wb")` actually gives me a binary mode file handle to write to. I don't get it, but that means the code is OK. I'll just fix the typo.